runtime/src/accounts_db.rs

```diff
@@ -816,40 +884,48 @@ impl AccountsDB {
         reclaims: SlotSlice<AccountInfo>,
         expected_single_dead_slot: Option<Slot>,
         no_dead_slot: bool,
-    ) {
-        if !reclaims.is_empty() {
-            let dead_slots = self.remove_dead_accounts(reclaims, expected_single_dead_slot);
-            if no_dead_slot {
-                assert!(dead_slots.is_empty());
-            } else if let Some(expected_single_dead_slot) = expected_single_dead_slot {
-                assert!(dead_slots.len() <= 1);
-                if dead_slots.len() == 1 {
-                    assert!(dead_slots.contains(&expected_single_dead_slot));
-                }
-            }
-            if !dead_slots.is_empty() {
-                self.process_dead_slots(&dead_slots);
+    ) -> (PurgedAccountSlots, ReclaimedAccounts) {
+        if reclaims.is_empty() {
+            return (HashMap::new(), HashMap::new());
+        }
+
+        let (dead_slots, removed_accounts) =
+            self.remove_dead_accounts(reclaims, expected_single_dead_slot);
+        if no_dead_slot {
+            assert!(dead_slots.is_empty());
+        } else if let Some(expected_single_dead_slot) = expected_single_dead_slot {
+            assert!(dead_slots.len() <= 1);
+            if dead_slots.len() == 1 {
+                assert!(dead_slots.contains(&expected_single_dead_slot));
             }
         }
+        let purged_account_slots = if !dead_slots.is_empty() {
+            self.process_dead_slots(&dead_slots)
+        } else {
+            HashMap::new()
+        };
```

Comment:
how about moving this code block into the callee like this?: https://github.com/solana-labs/solana/pull/12606/files#r498703041